### PR TITLE
Add support for by-name arguments

### PR DESCRIPTION
**Pull request by bbonanno**
Add support for Scala by-name arguments on methods.

It currently relies on a hack which is overriding ```org.mockito.internal.invocation.ArgumentProcessor```, ideally I'd prefer some functionality (maybe on MockSettings) so anyone can specify some pre-processing on the arguments without having to hack it like this.
I'd say this could be merged like this for now, but I wonder if the core mockito team can think in a better solution.

the extra behaviour is added on lines 27 and 30-32 of ```org.mockito.internal.invocation.ArgumentProcessor```

### CODE REVIEW

**Comment by mockitoguy** [comment]
Can you elaborate the use case? Please be patient, I don't work with Scala :)

**Comment by bbonanno** [comment]
@mockitoguy Sure!

So in Scala there is the concept of a by-name argument for a method, basically what it means is that the parameter is not evaluated until it is used for the first time inside the method that receives it.

A good example/comparison of this is when you have to log something, in Java you should not do something like
```java
logger.debug("some string " + some var + " another string")
```
and let's assume that method looks like 
```java
public void debug(String: txt) {
    if(Debug.isEnabled) {
         System.out.println(txt)
    }
}
```
if you do that, even if the 'debug' level is not enabled, every time your code executes the string concatenation will happen anyway, as it needs to be resolved so the method can be called.

in Scala, if you define that argument as lazy (by-name), then the string concatenation will not happen, because if the line inside the if is not called, then the argument is not evaluated...

Now, how does this work? basically the compiler, under the hood, will transform the method and the call to be something like this (this is not real code, just for example purposes)

```java
logger.debug(() -> "some string " + some var + " another string")

public void debug(Supplier<String>: txt) {
    if(Debug.isEnabled) {
         System.out.println(txt.get())
    }
}
```

So what happens there is that the lambda will only be evaluated if we go inside the if statement, so effectively, the string concatenation only happens if we are actually logging on debug mode
All of this is transparent in Scala and you never see that automagically generated lambda, you'd call the method as normal and the compiler will do the magic for you

Now, when we are in mockito/reflection land it is a different story, because all of this executes in Java land, Mockito gets a method that receives a Function and not the real argument, so for example, if we use ArgumentMatchers, it will get Functions that return ArgumentMatchers, and because of that it will fail saying that you didn't provide a matcher for that argument, something similar happens if you use normal values instead of ArgumentMatchers, because you pass normal values but the mock gets functions, the stubbings are not created properly and verify never works as the values provided are != to the function the mock gets

So basically, that's why I need this hook to force the evaluation of said Functions, and provide the mock with the real values

Makes sense?

**Comment by bbonanno** [comment]
@mockitoguy Let's not block the release on this, I found some bugs I still need to fix, but it would be nice if there was a less hacky way to do it

**Comment by mockitoguy** [comment]
Sounds good. I'm happy to work with you to solve this use case cleanly. Can you create a ticket to track it? The description of the use case is really good, thank you for elaborating!

**Comment by mockitoguy** [comment]
>So basically, that's why I need this hook to force the evaluation of said Functions, and provide the mock with the real values

My initial thoughts: we have 2 options:
 - create an explicit matcher method for that, e.g. verify(mock).foo(byName("bar"))
 - create an extension point. Options, in order of preference (API simplicity), best solution first:
   - If we control creation of mocks, we can pass some kind of argument process on mock creation.
   - If we cannot control the creation of mocks, but we control test execution, we can use the approach similar to MockitoSession.
   - If we cannot control mock creation, nor test execution, we can use classic plugin loaded from the classpath (example: MockMaker).

If we can add an extension point via mock creation it would be ideal.

**Comment by bbonanno** [comment]
@mockitoguy I tried option 1 before (a byName matcher), I couldn't make it work as the type of matcher is lost by the time we evaluate the arguments, so we don't know if that argument is effectively lazy.

But we definitively control the creation of mocks, so I'd go for an extension point (which is what I had in mind btw)


On a different matter, could we start releasing what we have so far? I'm quite keen to get people using it :)
